lib/web_ui/lib/src/engine/semantics/text_field.dart

```diff
@@ -341,75 +375,103 @@ class TextField extends RoleManager {
     }), true);
   }
 
-  bool _hasFocused = false;
-
-  @override
-  void update() {
-    // The user is editing the semantic text field directly, so there's no need
-    // to do any update here.
-    if (semanticsObject.hasLabel) {
-      editableElement.setAttribute(
-        'aria-label',
-        semanticsObject.label!,
-      );
-    } else {
-      editableElement.removeAttribute('aria-label');
-    }
-
-    editableElement.style
-      ..width = '${semanticsObject.rect!.width}px'
-      ..height = '${semanticsObject.rect!.height}px';
+  void _applyTextFromEditingState() {
+    final EditingState editingState = EditingState(
+      text: semanticsObject.value,
+      baseOffset: semanticsObject.textSelectionBase,
+      extentOffset: semanticsObject.textSelectionExtent,
+    );
+    editingState.applyTextToDomElement(editableElement);
+  }
 
-    // Whether we should request that the browser shift focus to the editable
-    // element, so that both the framework and the browser agree on what's
-    // currently focused.
-    bool needsDomFocusRequest = false;
+  void _setEditingState() {
     final EditingState editingState = EditingState(
       text: semanticsObject.value,
       baseOffset: semanticsObject.textSelectionBase,
       extentOffset: semanticsObject.textSelectionExtent,
     );
-    if (semanticsObject.hasFocus) {
-      if (!_hasFocused) {
-        _hasFocused = true;
-        SemanticsTextEditingStrategy.instance.activate(this);
-        needsDomFocusRequest = true;
-      }
-      if (domDocument.activeElement != editableElement) {
-        needsDomFocusRequest = true;
-      }
-      // Focused elements should have full text editing state applied.
-      SemanticsTextEditingStrategy.instance.setEditingState(editingState);
-    } else if (_hasFocused) {
+    SemanticsTextEditingStrategy.instance.setEditingState(editingState);
+  }
+
+  void _invokeIosWorkaround() {
+    if (editableElement != null) {
+      return;
+    }
+
+    _createEditableElement();
+    activeEditableElement.style.transform = 'translate(-999px,-999px)';
+    _positionInputElementTimer?.cancel();
+    _positionInputElementTimer = Timer(_delayBeforePlacement, () {
+      editableElement?.style.transform = '';
+    });
+
+    activeEditableElement.focus();
+    semanticsObject.element.removeAttribute('role');
```

Comment:
Is this to avoid having two elements with the "textfield" role representing the same text field? If so, might be worth leaving a comment explaining it for a future reader.

Reply:
Yes, VoiceOver starts acting weird if we don't remove the `role`.
Sure! I added some comments.